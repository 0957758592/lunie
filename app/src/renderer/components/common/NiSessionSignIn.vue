<template lang="pug">
.ni-session: form-struct(:submit='onSubmit').ni-session-container
  .ni-session-header
    .ni-session-title Sign In
    a(@click="help"): i.material-icons help_outline
  .ni-session-main
    form-group(:error='$v.fields.signInPassword.$error'
      field-id='sign-in-password' field-label='Password')
      field#sign-in-password(
        type="password"
        v-model="fields.signInPassword")
      form-msg(name='Password' type='required' v-if='!$v.fields.signInPassword.required')
      form-msg(name='Password' type='minLength' min="10" v-if='!$v.fields.signInPassword.minLength')
    form-group
      a(@click="setState('delete')") Sign Out and Remove Account
  .ni-session-footer
    btn(icon="arrow_forward" icon-pos="right" value="Next" size="lg")
</template>

<script>
import {required, minLength} from 'vuelidate/lib/validators'
import Btn from '@nylira/vue-button'
import Field from '@nylira/vue-field'
import FieldGroup from 'common/NiFieldGroup'
import FormGroup from 'common/NiFormGroup'
import FormMsg from '@nylira/vue-form-msg'
import FormStruct from 'common/NiFormStruct'
export default {
  name: 'ni-session-sign-in',
  components: {
    Btn,
    Field,
    FieldGroup,
    FormGroup,
    FormMsg,
    FormStruct
  },
  data: () => ({
    fields: {
      signInPassword: ''
    }
  }),
  methods: {
    help () { this.$store.commit('setModalHelp', true) },
    setState (value) { this.$store.commit('setModalSessionState', value) },
    async onSubmit () {
      this.$v.$touch()
      if (this.$v.$error) return
      try {
        await this.$store.dispatch('testLogin', {password: this.fields.signInPassword})
        this.$store.dispatch('signIn', {password: this.fields.signInPassword})
        this.$store.commit('setModalSession', false)
        this.$store.commit('notify', { title: 'Signed In', body: `You are now signed in to your Cosmos account.` })
      } catch (err) {
        this.$store.commit('notifyError', { title: 'Signing In Failed', body: err.message })
      }
    }
  },
  mounted () {
    this.$el.querySelector('#sign-in-password').focus()
  },
  validations: () => ({
    fields: {
      signInPassword: { required, minLength: minLength(10) }
    }
  })
}
</script>
<style lang="stylus">
  .ni-form-group
    a
      cursor pointer
</style>
