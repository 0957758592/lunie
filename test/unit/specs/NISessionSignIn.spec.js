import Vuex from 'vuex'
import Vuelidate from 'vuelidate'
import { mount, createLocalVue } from 'vue-test-utils'
import htmlBeautify from 'html-beautify'
import NiSessionSignIn from 'common/NiSessionSignIn'

const user = require('renderer/vuex/modules/user').default({})

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(Vuelidate)

describe('NiSessionSignIn', () => {
  let wrapper, store

  beforeEach(() => {
    store = new Vuex.Store({
      modules: {
        user
      }
    })
    wrapper = mount(NiSessionSignIn, {
      localVue,
      store
    })
    store.commit = jest.fn()
    store.dispatch = jest.fn(async () => null)
  })

  it('has the expected html structure', () => {
    expect(htmlBeautify(wrapper.html())).toMatchSnapshot()
  })

  it('should go back to the welcome screen on click', () => {
    wrapper.findAll('.ni-session-header a').at(0).trigger('click')
    expect(store.commit.mock.calls[0][0]).toBe('setModalSessionState')
    expect(store.commit.mock.calls[0][1]).toBe('welcome')
  })

  it('should open the help model on click', () => {
    wrapper.findAll('.ni-session-header a').at(1).trigger('click')
    expect(store.commit.mock.calls[0]).toEqual(['setModalHelp', true])
  })

  it('should close the modal on successful login', async () => {
    wrapper.setData({ fields: {
      signInPassword: '1234567890'
    }})
    await wrapper.vm.onSubmit()
    let calls = store.commit.mock.calls.map(args => args[0])
    expect(calls).toContain('setModalSession')
  })

  it('should signal signedin state on successful login', async () => {
    wrapper.setData({ fields: {
      signInPassword: '1234567890'
    }})
    await wrapper.vm.onSubmit()
    let calls = store.commit.mock.calls.map(args => args[0])
    expect(calls).toContain('notify')
    expect(calls).toContain('signIn')
  })

  it('should show error if password not 10 long', () => {
    wrapper.setData({ fields: {
      signInPassword: '123'
    }})
    wrapper.vm.onSubmit()
    expect(store.commit.mock.calls[0]).toBeUndefined()
    expect(wrapper.find('.ni-form-msg-error')).toBeDefined()
  })

  it('should show a notification if signin failed', async () => {
    store.dispatch = jest.fn(() => Promise.reject('Planed rejection'))
    wrapper.setData({ fields: {
      signInPassword: '1234567890'
    }})
    await wrapper.vm.onSubmit()
    expect(store.commit).toHaveBeenCalled()
    expect(store.commit.mock.calls[0][0]).toBe('notifyError')
  })
})
