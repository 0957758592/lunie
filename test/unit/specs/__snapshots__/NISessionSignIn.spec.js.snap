// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NISessionSignIn has the expected html structure 1`] = `"<div class=\\"ni-session\\"><form class=\\"ni-session-container ni-form\\"><div class=\\"ni-form-container\\"> <!----> <main class=\\"ni-form-main\\"><div class=\\"ni-session-header\\"><a><i class=\\"material-icons\\">arrow_back</i></a><div class=\\"ni-session-title\\">Enter Seed</div><a><i class=\\"material-icons\\">help_outline</i></a></div><div class=\\"ni-session-main\\"><div class=\\"ni-form-group\\"><label for=\\"sign-in-seed\\">Account Seed</label><div class=\\"ni-form-group-field\\"><textarea placeholder=\\"Enter your 12-word seed here\\" class=\\"ni-field\\" id=\\"sign-in-seed\\"></textarea><div class=\\"ni-form-msg ni-form-msg-error\\">Account Seed is required</div></div></div></div><div class=\\"ni-session-footer\\"><button class=\\"ni-btn\\"><span class=\\"ni-btn-container ni-btn-size-lg\\"><i aria-hidden=\\"true\\" class=\\"ni-btn-icon material-icons\\">exit_to_app</i><span class=\\"ni-btn-value\\">Sign In</span></span></button></div></main> <!----></div></form></div>"`;
